cpp/src/text/utilities/tokenize_ops.cuh

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/string_view.cuh>
+
+#include <thrust/logical.h>
+
+namespace cudf
+{
+namespace nvtext
+{
+namespace detail
+{
+
+using string_index_pair = thrust::pair<const char*,size_type>;
+
+/**
+ * @brief Base class for tokenize functions that use multi-character
+ * delimiters.
+ *
+ * This is common code for tokenize, token-counters, normalize functions.
+ * If an empty delimiter string is specified, then whitespace
+ * (code-point <= ' ') is used to identify tokens.
+ */
+struct base_tokenator
+{
+    string_view d_delimiter{}; // zero or more delimiter characters
+
+    /**
+     * @brief Return true if the given character is a delimiter.
+     *
+     * For empty delimiter, whitespace code-point is checked.
+     *
+     * @param chr The character to test.
+     * @return true if the character is a delimiter
+     */
+    __device__ bool is_delimiter(char_utf8 chr)
+    {
+        return d_delimiter.empty() ? (chr <= ' ') : // whitespace check
+               thrust::any_of( thrust::seq, d_delimiter.begin(), d_delimiter.end(),
+                               [chr] __device__ (char_utf8 c) {return c==chr;});
+    }
+
+    /**
+     * @brief Identifies the bounds of the next token in the given
+     * string at the specified iterator position.
+     *
+     * For empty delimiter, whitespace code-point is checked.
+     * Starting at the given iterator (itr) position, a token
+     * start position (spos) is identified when a delimiter is
+     * not found. Once found, the end position (epos) is identified
+     * when a delimiter or the end of the string is found.
+     *
+     * @param d_str String to tokenize.
+     * @param[in,out] spaces Identifies current character position
+     *                is a delimiter or not.
+     * @param[in,out] itr Current position in d_str to search for tokens.
+     * @param[in,out] spos Start character position of the token found.
+     * @param[in,out] epos End character position of the token found.
+     * @return true if a token was found, false if no more tokens
+     */
+    __device__ bool next_token( string_view const& d_str, bool& spaces,
```

Comment:
Would it be possible to write this in a way that doesn't require in/out parameters? This is kind of a confusing implementation as a result.

Reply:
Yes, I was thinking about turning these parameters into a struct that holds the tokenizing state. I think that could make the code a bit cleaner.